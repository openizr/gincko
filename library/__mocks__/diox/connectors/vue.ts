/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/**
 * diox/connectors/vue mock.
 */
const connect = vi.fn(() => vi.fn(() => ({
  steps: [{ id: 'start' }, { id: 'end' }],
  loading: process.env.LOADING === 'true',
})));

export default connect;
