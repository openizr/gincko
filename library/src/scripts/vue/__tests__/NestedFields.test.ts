/**
 * Copyright (c) Openizr. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @jest-environment jsdom
 */

import { render, waitFor } from '@testing-library/vue';
import NestedFields from 'scripts/vue/NestedFields.vue';

jest.mock('biuty/vue');
jest.mock('scripts/vue/FormField.vue');

describe('vue/NestedFields', () => {
  const onUserAction = jest.fn();
  const userInputs = { test: 'ok' };
  const variables = { var: 'value' };
  const i18n = jest.fn(() => 'TRANSLATED LABEL');

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders correctly - array', async () => {
    const { container } = render(NestedFields, {
      props: {
        t: i18n,
        type: 'array',
        label: 'Test',
        helper: 'Helper',
        path: 'path.0.to.field',
        variables,
        userInputs,
        fields: [{
          id: 'Message',
          status: 'initial',
          component: 'Message',
          componentProps: {},
        }, null],
        onUserAction,
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - array with minItems', async () => {
    const { container, rerender } = render(NestedFields, {
      props: {
        t: i18n,
        type: 'array',
        minItems: 2,
        path: 'path.0.to.field',
        variables,
        userInputs,
        fields: [{
          id: 'Message',
          status: 'initial',
          component: 'Message',
          componentProps: {},
        }, {
          id: 'Message',
          status: 'initial',
          component: 'Message',
          componentProps: {},
        }, {
          id: 'Message',
          status: 'initial',
          component: 'Message',
          componentProps: {},
        }],
        onUserAction,
      },
    });
    await waitFor(() => new Promise<void>((resolve) => { setTimeout(resolve, 50); }));
    expect(container.firstChild).toMatchSnapshot();
    await rerender({ minItems: 4 });
    expect(container.firstChild).toMatchSnapshot();
    await rerender({ minItems: undefined });
  });

  test('renders correctly - array with maxItems', async () => {
    const { container, rerender } = render(NestedFields, {
      props: {
        t: i18n,
        type: 'array',
        path: 'path.0.to.field',
        variables,
        userInputs,
        maxItems: 1,
        value: ['test', 'test'],
        fields: [{
          id: 'Message',
          status: 'initial',
          component: 'Message',
          componentProps: {},
        }, null],
        onUserAction,
      },
    });
    await waitFor(() => new Promise<void>((resolve) => { setTimeout(resolve, 50); }));
    expect(container.firstChild).toMatchSnapshot();
    await rerender({ maxItems: 0 });
    expect(container.firstChild).toMatchSnapshot();
    await rerender({ maxItems: undefined });
  });

  test('renders correctly - object', async () => {
    const { container } = render(NestedFields, {
      props: {
        t: i18n,
        type: 'object',
        path: 'path.0.to.field',
        variables,
        userInputs,
        fields: [{
          id: 'Message',
          status: 'initial',
          component: 'Message',
          componentProps: {},
        }, null],
        onUserAction,
      },
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - dynamicObject with valid pattern', async () => {
    const { container } = render(NestedFields, {
      props: {
        t: i18n,
        type: 'dynamicObject',
        path: 'path.0.to.field',
        variables,
        userInputs,
        fields: [{
          id: 'Message',
          status: 'initial',
          component: 'Message',
          componentProps: {},
        }, null],
        onUserAction,
        allowedPatterns: [/test/],
        addTextfieldProps: { debounceTimeout: 10 },
      },
    });
    await waitFor(() => new Promise<void>((resolve) => { setTimeout(resolve, 50); }));
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - dynamicObject with invalid pattern', async () => {
    const { container } = render(NestedFields, {
      props: {
        t: i18n,
        type: 'dynamicObject',
        path: 'path.0.to.field',
        variables,
        userInputs,
        fields: [{
          id: 'Message',
          status: 'initial',
          component: 'Message',
          componentProps: {},
        }, null],
        onUserAction,
        addTextfieldProps: { debounceTimeout: 10, helper: 'INVALID' },
      },
    });
    await waitFor(() => new Promise<void>((resolve) => { setTimeout(resolve, 50); }));
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders correctly - dynamicObject with error', async () => {
    const { container, rerender } = render(NestedFields, {
      props: {
        t: i18n,
        helper: 'Error',
        modifiers: 'error',
        type: 'dynamicObject',
        variables,
        path: 'path.0.to.field',
        userInputs,
        fields: [{
          id: 'Message',
          status: 'initial',
          component: 'Message',
          componentProps: {},
        }, {
          id: 'Message',
          status: 'initial',
          component: 'Message',
          componentProps: {},
        }, null],
        onUserAction,
      },
    });
    await rerender({
      value: {
        test: 'Test',
        test2: 'Test',
      },
    });
    await waitFor(() => new Promise<void>((resolve) => { setTimeout(resolve, 50); }));
    expect(container.firstChild).toMatchSnapshot();
  });
});
