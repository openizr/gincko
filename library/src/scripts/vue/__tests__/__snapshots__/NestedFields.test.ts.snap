// Vitest Snapshot v1

exports[`vue/NestedFields > renders correctly - array 1`] = `
<div
  class="ui-nested-fields  array"
>
  <span
    class="ui-nested-fields__label"
  >
    Test
  </span>
  
  <div
    class="ui-nested-fields__field"
  >
    <!--v-if-->
    <div
      type="button"
    >
      {}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  
  <div
    class="ui-nested-fields__add"
  >
    <!--v-if-->
    <div
      modifiers=" "
      type="button"
    >
      {}
    </div>
  </div>
  <span
    class="ui-nested-fields__helper"
  >
    Helper
  </span>
</div>
`;

exports[`vue/NestedFields > renders correctly - array with maxItems 1`] = `
<div
  class="ui-nested-fields  array"
>
  <!--v-if-->
  
  <div
    class="ui-nested-fields__field"
  >
    <!--v-if-->
    <div
      type="button"
    >
      {}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  
  <div
    class="ui-nested-fields__add"
  >
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`vue/NestedFields > renders correctly - array with maxItems 2`] = `
<div
  class="ui-nested-fields  array"
>
  <!--v-if-->
  
  <div
    class="ui-nested-fields__field"
  >
    <!--v-if-->
    <div
      type="button"
    >
      {}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  
  <div
    class="ui-nested-fields__add"
  >
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
`;

exports[`vue/NestedFields > renders correctly - array with minItems 1`] = `
<div
  class="ui-nested-fields  array"
>
  <!--v-if-->
  
  <div
    class="ui-nested-fields__field"
  >
    <!--v-if-->
    <div
      type="button"
    >
      {}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  <div
    class="ui-nested-fields__field"
  >
    <!--v-if-->
    <div
      type="button"
    >
      {}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  <div
    class="ui-nested-fields__field"
  >
    <!--v-if-->
    <div
      type="button"
    >
      {}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  
  <div
    class="ui-nested-fields__add"
  >
    <!--v-if-->
    <div
      modifiers=" "
      type="button"
    >
      {}
    </div>
  </div>
  <!--v-if-->
</div>
`;

exports[`vue/NestedFields > renders correctly - array with minItems 2`] = `
<div
  class="ui-nested-fields  array"
>
  <!--v-if-->
  
  <div
    class="ui-nested-fields__field"
  >
    <!--v-if-->
    <!--v-if-->
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  <div
    class="ui-nested-fields__field"
  >
    <!--v-if-->
    <!--v-if-->
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  <div
    class="ui-nested-fields__field"
  >
    <!--v-if-->
    <!--v-if-->
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  
  <div
    class="ui-nested-fields__add"
  >
    <!--v-if-->
    <div
      modifiers=" "
      type="button"
    >
      {}
    </div>
  </div>
  <!--v-if-->
</div>
`;

exports[`vue/NestedFields > renders correctly - dynamicObject with error 1`] = `
<div
  class="ui-nested-fields error dynamicObject"
>
  <!--v-if-->
  
  <div
    class="ui-nested-fields__field"
  >
    <span
      class="ui-nested-fields__field__label"
    >
      Message
    </span>
    <div
      type="button"
    >
      {}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  <div
    class="ui-nested-fields__field"
  >
    <span
      class="ui-nested-fields__field__label"
    >
      Message
    </span>
    <div
      type="button"
    >
      {}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  
  <div
    class="ui-nested-fields__add"
  >
    <div
      debounce-timeout="100"
      name="undefined-add"
      readonly=""
      value=""
    >
      {}
    </div>
    <div
      modifiers="disabled "
      type="button"
    >
      {}
    </div>
  </div>
  <span
    class="ui-nested-fields__helper"
  >
    Error
  </span>
</div>
`;

exports[`vue/NestedFields > renders correctly - dynamicObject with invalid pattern 1`] = `
<div
  class="ui-nested-fields  dynamicObject"
>
  <!--v-if-->
  
  <div
    class="ui-nested-fields__field"
  >
    <span
      class="ui-nested-fields__field__label"
    >
      Message
    </span>
    <div
      type="button"
    >
      {}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  
  <div
    class="ui-nested-fields__add"
  >
    <div
      debounce-timeout="10"
      helper="INVALID"
      name="undefined-add"
      readonly=""
      value=""
    >
      {}
    </div>
    <div
      modifiers="disabled "
      type="button"
    >
      {}
    </div>
  </div>
  <!--v-if-->
</div>
`;

exports[`vue/NestedFields > renders correctly - dynamicObject with valid pattern 1`] = `
<div
  class="ui-nested-fields  dynamicObject"
>
  <!--v-if-->
  
  <div
    class="ui-nested-fields__field"
  >
    <span
      class="ui-nested-fields__field__label"
    >
      Message
    </span>
    <div
      type="button"
    >
      {}
    </div>
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  
  <div
    class="ui-nested-fields__add"
  >
    <div
      debounce-timeout="10"
      name="undefined-add"
      readonly=""
      value=""
    >
      {}
    </div>
    <div
      modifiers="disabled "
      type="button"
    >
      {}
    </div>
  </div>
  <!--v-if-->
</div>
`;

exports[`vue/NestedFields > renders correctly - object 1`] = `
<div
  class="ui-nested-fields  object"
>
  <!--v-if-->
  
  <div
    class="ui-nested-fields__field"
  >
    <!--v-if-->
    <!--v-if-->
    <div
      id="field"
    >
      {"field":{"id":"Message","status":"initial","component":"Message","componentProps":{}},"path":"path.0.to.field.Message","isActive":false,"variables":{"var":"value"},"userInputs":{"test":"ok"}}
    </div>
  </div>
  
  <!--v-if-->
  <!--v-if-->
</div>
`;
